presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -482,7 +484,8 @@ public PlanWithProperties visitTopNRowNumber(TopNRowNumberNode node, StreamPrefe
         @Override
         public PlanWithProperties visitTableWriter(TableWriterNode originalTableWriterNode, StreamPreferredProperties parentPreferences)
         {
-            if (getTaskWriterCount(session) == 1) {
+            if (getTaskWriterCount(session) == 1
+                    || (originalTableWriterNode.getPartitioningScheme().isPresent() && !isConcurrentWritesToPartitionedTableEnabled(session))) {
```

Comment:
Maybe I miss something but I don't see why we need this "or" condition.

Reply:
This is just a guard flag. If `isConcurrentWritesToPartitionedTableEnabled` returns `false` - it will shortcut to a gathering exchange before the `TableWriterOperator` vs the partitioned exchange on the bucket function.